File: sdk/serviceability/python/serviceability/pda.py
"""PDA derivation for serviceability program accounts."""

from solders.pubkey import Pubkey  # type: ignore[import-untyped]

SEED_PREFIX = b"doublezero"
SEED_GLOBAL_STATE = b"globalstate"
SEED_GLOBAL_CONFIG = b"config"
SEED_PROGRAM_CONFIG = b"programconfig"
SEED_TENANT = b"tenant"


def derive_global_state_pda(program_id: Pubkey) -> tuple[Pubkey, int]:
    return Pubkey.find_program_address([SEED_PREFIX, SEED_GLOBAL_STATE], program_id)


def derive_global_config_pda(program_id: Pubkey) -> tuple[Pubkey, int]:
    return Pubkey.find_program_address([SEED_PREFIX, SEED_GLOBAL_CONFIG], program_id)


def derive_program_config_pda(program_id: Pubkey) -> tuple[Pubkey, int]:
    return Pubkey.find_program_address([SEED_PREFIX, SEED_PROGRAM_CONFIG], program_id)


def derive_tenant_pda(program_id: Pubkey, code: str) -> tuple[Pubkey, int]:
    return Pubkey.find_program_address([SEED_PREFIX, SEED_TENANT, code.encode()], program_id)

File: sdk/serviceability/typescript/serviceability/pda.ts
import { PublicKey } from "@solana/web3.js";

const SEED_PREFIX = Buffer.from("doublezero");
const SEED_GLOBAL_STATE = Buffer.from("globalstate");
const SEED_GLOBAL_CONFIG = Buffer.from("config");
const SEED_PROGRAM_CONFIG = Buffer.from("programconfig");
const SEED_TENANT = Buffer.from("tenant");

export function deriveGlobalStatePda(
  programId: PublicKey,
): [PublicKey, number] {
  return PublicKey.findProgramAddressSync([SEED_PREFIX, SEED_GLOBAL_STATE], programId);
}

export function deriveGlobalConfigPda(
  programId: PublicKey,
): [PublicKey, number] {
  return PublicKey.findProgramAddressSync([SEED_PREFIX, SEED_GLOBAL_CONFIG], programId);
}

export function deriveProgramConfigPda(
  programId: PublicKey,
): [PublicKey, number] {
  return PublicKey.findProgramAddressSync([SEED_PREFIX, SEED_PROGRAM_CONFIG], programId);
}

export function deriveTenantPda(
  programId: PublicKey,
  code: string,
): [PublicKey, number] {
  return PublicKey.findProgramAddressSync(
    [SEED_PREFIX, SEED_TENANT, Buffer.from(code)],
    programId,
  );
}
